### PR TITLE
ext_authz: Add per-route flag to skip request body buffering 

### LINKED ISSUE #13285: Support for per route "with_request_body" configuration for ext_authz
*Title*: Support for per route "with_request_body" configuration for ext_authz

*Description*:
For the ext_authz extension we would like to be able to configure with_request_body for specific routes. Our use case is for enabling SAML from our ext_authz service which requires posting the SAML assertion in the body for authentication to a specific endpoint. We would like to be able to do this without the additional overhead of passing the body on every request other request.

### PR DESCRIPTION
**Pull request by dio**
Commit Message: This patch adds per-route flag to bypass the request body buffering.
Additional Description: I'm preparing another patch to skip request body buffering via a list of `config.route.v3.HeaderMatcher`. Hence skipping request body buffering, for example, can be done to a specific request with a certain "content-type".

Risk Level: Low
Testing: Added.
Docs Changes: Updated.
Release Notes: Added.
Fixes #13285

Signed-off-by: Dhi Aurrahman <dio@tetrate.io>

### CODE REVIEW

File: source/extensions/filters/http/ext_authz/ext_authz.h [hunk 1]
```diff
@@ -269,6 +275,7 @@ class Filter : public Logger::Loggable<Logger::Id::filter>,
   bool buffer_data_{};
   bool skip_check_{false};
   envoy::service::auth::v3::CheckRequest check_request_{};
+  mutable const FilterConfigPerRoute* per_route_config_{nullptr};
```
Comment (mattklein123) [inline source/extensions/filters/http/ext_authz/ext_authz.h:278]: It seems odd that this is mutable. Why does it need to be mutable?

Also, technically, I think it's possible for this to be unsafe if you store a raw pointer like this. I think you need to also snap the RoutePtr and hold the shared_ptr to make sure the per route config doesn't go away underneath you?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Generally LGTM with a small comment. Thanks!

/wait

File: source/extensions/filters/http/ext_authz/ext_authz.h (hunk 1 above)
Comment (dio) [reply source/extensions/filters/http/ext_authz/ext_authz.h]: Yes, this is not necessary. I introduced a struct instead when getting the specific config, and hold those flags to the existing (member) booleans.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**